File: reactapp/src/services/complaintService.js
import api from "./api";

export const complaintService = {
  getAllComplaints: async () => {
    const response = await api.get("/complaints");
    return Array.isArray(response.data) ? response.data : [];
  },

  getComplaintById: async (id) => {
    const response = await api.get(`/complaints/${id}`);
    return response.data;
  },

  createComplaint: async (complaintData) => {
    const response = await api.post("/complaints", complaintData);
    return response.data;
  },

  assignComplaint: async (complaintId, employeeId) => {
    const response = await api.put(`/complaints/${complaintId}/assign/${employeeId}`);
    return response.data;
  },

  updateStatus: async (complaintId, status) => {
    const response = await api.put(`/complaints/${complaintId}/status`, null, {
      params: { status }
    });
    return response.data;
  }
};

export default complaintService;

File: springapp/src/main/java/com/examly/springapp/service/ComplaintService.java
package com.examly.springapp.service;

import com.examly.springapp.dto.complaint.ComplaintRequest;
import com.examly.springapp.model.Complaint;

import java.util.List;

public interface ComplaintService {

    Complaint createComplaint(
            ComplaintRequest request,
            String userEmail
    );

    List<Complaint> getAllComplaints();

    Complaint getComplaintById(Long id);

    Complaint assignComplaint(
            Long complaintId,
            Long employeeId
    );

    Complaint updateStatus(
            Long complaintId,
            String status
    );

    Complaint updateComplaint(
            Long complaintId,
            com.examly.springapp.dto.complaint.ComplaintUpdateRequest request,
            String userEmail,
            boolean isAdmin
    );

    void deleteComplaint(
            Long complaintId,
            String userEmail,
            boolean isAdmin
    );

    List<com.examly.springapp.model.ComplaintUpdate> getComplaintHistory(Long complaintId);
}